src/main/scala/org/ergoplatform/nodeView/ErgoNodeViewHolder.scala

```diff
@@ -408,12 +408,22 @@ abstract class ErgoNodeViewHolder[State <: ErgoState[State]](settings: ErgoSetti
               case Success(newMinState) =>
                 val newMemPool = updateMemPool(progressInfo.toRemove, blocksApplied, memoryPool(), newMinState)
 
-                //we consider that vault always able to perform a rollback needed
                 @SuppressWarnings(Array("org.wartremover.warts.OptionPartial"))
                 val newVault = if (progressInfo.chainSwitchingNeeded) {
-                  vault().rollback(idToVersion(progressInfo.branchPoint.get)).get
-                } else vault()
-                blocksApplied.foreach(newVault.scanPersistent)
+                  val v = vault()
+                  v.rollback(idToVersion(progressInfo.branchPoint.get)) match {
+                    case Success(nv) => nv
+                    case Failure(e) => log.warn("Wallet rollback failed: ", e); v
+                  }
+                } else {
+                  vault()
+                }
+
+                val almostSyncedGap = 20
+
+                if((newHistory.headersHeight - newHistory.fullBlockHeight) < almostSyncedGap) {
+                  blocksApplied.foreach(newVault.scanPersistent)
```

Comment:
but does not it mean that the previous blocks will be missing in Vault? In comparison to previous way

Reply:
No, the wallet will automatically scan previous blocks by sending ScanInThePast messages to self